### PR TITLE
Reduce memory footprint

### PR DESCRIPTION
**Pull request by guykhazma**
### What changes are proposed in this pull request?

- Save the partition values as part of the metadata in order to get indexed and required files only for the relevant partitions
- Clear the state of the DataSkippingFileFilter after handling stats.
This should help to reduce the memory foot print as we don't need the `required` and `indexed` files once skipping was finished.
- Get all not required files and keep them as the state instead of having both the indexed files and required files

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

- Adding tests for all partition data types
- Running Existing tests

### CODE REVIEW

File: src/main/scala/io/xskipper/index/execution/MetadataProcessor.scala [hunk 1]
```diff
@@ -292,39 +314,45 @@ class MetadataProcessor(spark: SparkSession, uri: String, metadataHandle: Metada
...
     // indexed in the metadatastore)
-    val newOrModifiedFileIds = fileIds.filter {
-      case (fileLocation: String, fileID: String) =>
-        !allIndexedFiles.contains(fileID)
-    }
-
-    // getting the list of fileIDs to be removed since the object has been updated
-    // or deleted since last indexing
-    val removedFilesIds = allIndexedFiles.diff(fileIds.unzip._2.toSet)
+    val newOrModifiedPartitionDirectories = partitionDirectories.flatMap(pd => {
```
Comment (guykhazma) [inline src/main/scala/io/xskipper/index/execution/MetadataProcessor.scala:339]: This is done to avoid looping twice over the `allIndexedFiles`

File: src/main/scala/io/xskipper/search/DataSkippingFileFilter.scala [hunk 2]
```diff
@@ -207,5 +219,19 @@ class DataSkippingFileFilter(tid: String,
 
     // clear current filter statistics so next run will start with empty stats
     clearCurrentFilterStatistics()
+    // reset the file filter state to force cleanup of memory
+    resetState()
```
Comment (guykhazma) [inline src/main/scala/io/xskipper/search/DataSkippingFileFilter.scala:223]: this is done to force freeing up the memory once skipping is finished

File: src/main/scala/io/xskipper/metadatastore/parquet/ParquetMetadataHandle.scala [hunk 3]
```diff
@@ -466,7 +501,7 @@ class ParquetMetadataHandle(val session: SparkSession, tableIdentifier: String)
     val baseDf = getMetaDataDFRaw(false)
 
     val indexes = getIndexes(Registration.getCurrentIndexFactories())
-    ParquetUtils.getTransformedDataFrame(baseDf, indexes, false)
+    ParquetUtils.getTransformedDataFrame(baseDf, indexes, None, false)
```
Comment (guykhazma) [inline src/main/scala/io/xskipper/metadatastore/parquet/ParquetMetadataHandle.scala:504]: is this fine?

File: src/main/scala/io/xskipper/metadatastore/parquet/ParquetUtils.scala [hunk 4]
```diff
@@ -280,7 +291,7 @@ object ParquetUtils extends Logging {
     // this only works since we need the schema tree (names and types) and not metadata, so
     // we can live with the incorrect md returned in the schema
     val parquetSchema = converter.convert(
-      createDFSchema(indexes, includeExtraColumns = false, "", None))
+      createDFSchema(indexes, None, includeExtraColumns = false, "", None))
```
Comment (guykhazma) [inline src/main/scala/io/xskipper/metadatastore/parquet/ParquetUtils.scala:294]: is this fine?

File: src/main/scala/io/xskipper/metadatastore/parquet/ParquetUtils.scala [hunk 5]
```diff
@@ -206,8 +216,9 @@ object ParquetUtils extends Logging {
     // if encryption is ON, then the footer key must be set regardless
     // of plaintext footer being set or not (used for footer signing by PME
     // even if the footer itself is left plaintext).
-    // in that case, if plaintext footer is set to ON (default) then the footer
-    // and the obj_name column are both encrypted with the footer key.
+    // in that case, if plaintext footer is set to OFF (default) then the footer
+    // and the obj_name column are both encrypted with the footer key. the partition (hive style)
+    // columns are also encrypted using the footer key.
```
[the remark this answers is not in the document]
Comment (gallushi) [reply src/main/scala/io/xskipper/metadatastore/parquet/ParquetUtils.scala:221]: yes

File: src/main/scala/io/xskipper/metadatastore/parquet/ParquetUtils.scala (hunk 4 above)
Comment (gallushi) [reply src/main/scala/io/xskipper/metadatastore/parquet/ParquetUtils.scala:294]: yes, updated the comment that explains this.
TL;DR: this call is only to retrieve the column paths for a specific index, and is not influenced by the partition schema in any way - meaning in particular it can be `None`.

**Comment by guykhazma** [comment]
merged in #60
